### PR TITLE
Cartesian indexing for SubArrays with non-integer AbstractRanges

### PR DESCRIPTION
**Pull request by jishnub**
With #50457 now merged, an `AbstractRange` of `CartesianIndex`es should use `CartesianIndexing` in constructing a `SubArray`. I've limited the method to integer ranges, as this is the case where we know for sure that the indexing may be linear.

Fixes
```julia
julia> view(1:2, StepRangeLen(CartesianIndex(1), CartesianIndex(1), 0))
0-element view(::UnitRange{Int64}, StepRangeLen(CartesianIndex(1,), CartesianIndex(1,), 0)) with eltype Int64
```

### CODE REVIEW

**Comment by Seelengrab** [comment]
Does this have an effect on https://github.com/JuliaLang/julia/issues/43336?

**Comment by jishnub** [comment]
I doubt that's related, this is about using `CartesianIndexing` instead of `LinearIndexing` when a range of `CartesianIndex`es is used in constructing a view. Do you see a connection?

File: base/subarray.jl [hunk 1]
```diff
@@ -53,7 +53,7 @@ viewindexing(I::Tuple{Slice, Slice, Vararg{Any}}) = (@inline; viewindexing(tail(
 viewindexing(I::Tuple{Slice, AbstractUnitRange, Vararg{ScalarIndex}}) = IndexLinear()
 viewindexing(I::Tuple{Slice, Slice, Vararg{ScalarIndex}}) = IndexLinear() # disambiguate
 # In general, ranges are only fast if all other indices are scalar
-viewindexing(I::Tuple{AbstractRange, Vararg{ScalarIndex}}) = IndexLinear()
+viewindexing(I::Tuple{AbstractRange{<:Integer}, Vararg{ScalarIndex}}) = IndexLinear()
```
Comment (N5N3) [inline base/subarray.jl:56]:
```suggestion
viewindexing(I::Tuple{AbstractRange{<:ScalarIndex}, Vararg{ScalarIndex}}) = IndexLinear()
```
LGTM, perhaps it keeps more consistency if we use `ScalarIndex` here.
And the comment above could also be updated?

**Comment by jishnub** [comment]
Good to go?

**Pull request merged**